components/metris/internal/provider/azure/azure.go

```diff
@@ -103,93 +106,130 @@ func (a *Azure) Run(ctx context.Context) {
-						case http.StatusNotFound:
-							if strings.Contains(errdetail.Original.Error(), responseErrCodeResourceGroupNotFound) {
-								instance.retryAttempts++
-
-								if instance.retryAttempts < maxRetryAttempts {
-									a.instanceStorage.Put(instance.cluster.TechnicalID, instance)
-									workerlogger.Warnf("can't find resource group in azure, attempts: %d/%d", instance.retryAttempts, maxRetryAttempts)
-								} else {
-									a.instanceStorage.Delete(instance.cluster.TechnicalID)
-									workerlogger.Warnf("removing cluster after %d attempts", maxRetryAttempts)
-								}
-							}
-
-						case http.StatusTooManyRequests:
-							// request is being throttled, skip next call to release pressure on API
-							instance.retryBackoff = true
-
-							workerlogger.Debug("=============> THROTTLING - setting retryBackoff")
-						}
-					}
+func (a *Azure) processInstance(workerlogger log.Logger, instance *Instance, ctx context.Context, getMetrics metricsGetter) bool {
+	if tracing.IsEnabled() {
+		var span *trace.Span
+		ctx, span = trace.StartSpan(ctx, "metris/provider/azure/processInstance")
+		defer span.End()
 
-					if instance.lastEvent == nil {
-						workerlogger.With("error", err).Error("could not get metrics, dropping events because no cached information")
-					} else {
-						workerlogger.With("error", err).Error("could not get metrics, using information from cache")
+		workerlogger = workerlogger.With("traceID", span.SpanContext().TraceID).With("spanID", span.SpanContext().SpanID)
+	}
 
-						eventData = instance.lastEvent
-					}
-				}
+	vmcaps := make(vmCapabilities)
 
-				if eventData != nil {
-					if err := a.sendMetrics(workerlogger, instance, eventData); err != nil {
-						workerlogger.With("error", err).Error("error parsing metric information, could not send event to EDP")
-					}
-				}
+	if obj, exists := a.vmCapsStorage.Get(instance.cluster.Region); exists {
+		if caps, ok := obj.(*vmCapabilities); ok {
+			vmcaps = *caps
+		}
+	} else {
+		workerlogger.Warnf("vm capabilities for region %s not found, some metrics won't be available", instance.cluster.Region)
+	}
 
-				// save changes to storage
-				a.instanceStorage.Put(instance.cluster.TechnicalID, instance)
+	var (
+		eventData       *EventData
+		err             error
+		errorOccurred   bool
+		rateLimited     bool
+		instanceDeleted bool
+	)
 
-				a.queue.Done(clusterid)
+	eventData, err = getMetrics(ctx, workerlogger, instance, &vmcaps, a.config.PollInterval, a.config.PollingDuration)
+
+	if err != nil {
+		errorOccurred = true
+		eventData, rateLimited, instanceDeleted = processError(ctx, workerlogger, instance, eventData, err, a.config.MaxRetries, a.instanceStorage)
+	}
 
-				// requeue item after X duration if client still in storage
-				if !a.queue.ShuttingDown() {
-					workerlogger.Debugf("requeuing cluster in %s", a.config.PollInterval)
-					a.queue.AddAfter(clusterid, a.config.PollInterval)
+	if !errorOccurred {
```

Comment:
Is this "if" really necessary, could it be merge with the previous error check? I don't see the variable "errorOccurred" used anywhere else.

Reply:
done